File: computercraft/try_move.py
from typing import Callable, Tuple

from .try_clean_inventory import try_clean_inventory
from .constants import Mock

turtle = Mock.Turtle()


def try_move(move: str, is_reverse_order: bool = False) -> bool:
    """
    Reverse order only affects whether turns are executed before movement. Any reversal of direction
    should be carried out externally with the resulting direction being used in this function
    """

    def apply_turns(is_reverse: bool = False) -> None:
        if not is_reverse:
            if move == "left":
                turtle.turnLeft()
            elif move == "right":
                turtle.turnRight()
            elif move == "back":
                turtle.turnRight()
                turtle.turnRight()

        else:
            if move == "left":
                turtle.turnRight()
            elif move == "right":
                turtle.turnLeft()
            elif move == "back":
                turtle.turnLeft()
                turtle.turnLeft()

    def get_commands() -> Tuple[Callable, Callable]:
        if move == "up":
            return turtle.up, turtle.digUp
        elif move == "down":
            return turtle.down, turtle.digDown
        else:
            return turtle.forward, turtle.dig

    def try_move_command() -> bool:
        for attempt in range(100):  # and i would try 100 times and i would try 100 more
            if move_command():
                return True
            if attempt == 99:
                return False

            if not try_clean_inventory():
                return False
            dig_command()

    if move not in {"up", "down", "left", "right", "forward", "back"}:
        raise ValueError(move)

    move_command, dig_command = get_commands()

    if not is_reverse_order:
        apply_turns()
        if try_move_command():
            return True
        else:
            apply_turns(is_reverse=True)
            return False

    else:
        if try_move_command():
            apply_turns()
            return True
        else:
            return False

File: computercraft/try_move.lua
local try_clean_inventory = require("try_clean_inventory.lua")


function try_move(move, is_reverse_order)
    --[[
    Reverse order only affects whether turns are executed before movement. Any reversal of direction
    should be carried out externally with the resulting direction being used in this function
    --]]

    function apply_turns(is_reverse)
        if not is_reverse then
            if move == "left" then
                turtle.turnLeft()
            elseif move == "right" then
                turtle.turnRight()
            elseif move == "back" then
                turtle.turnRight()
                turtle.turnRight()
            end

        else
            if move == "left" then
                turtle.turnRight()
            elseif move == "right" then
                turtle.turnLeft()
            elseif move == "back" then
                turtle.turnLeft()
                turtle.turnLeft()
            end
        end
    end

    function get_commands()
        if move == "up" then
            return {[1] = turtle.up, [2] = turtle.digUp}
        elseif move == "down" then
            return {[1] = turtle.down, [2] = turtle.digDown}
        else
            return {[1] = turtle.forward, [2] = turtle.dig}
        end
    end

    function try_move_command()
        for attempt=1,100 do
            if move_command() then
                return true
            elseif attempt == 100 then
                return false
            elseif not try_clean_inventory() then
                return false
            end

            dig_command()
        end
    end

    local valid_moves = {up=true, down=true, left=true, right=true, forward=true, back=true}
    if valid_moves[move] == nil then
        error(move)
    end

    local commands = get_commands()
    local move_command = commands[1]
    local dig_command = commands[2]

    if not is_reverse_order then
        apply_turns()
        if try_move_command() then
            return true
        else
            apply_turns(true)
            return false
        end

    else
        if try_move_command() then
            apply_turns()
            return true
        else
            return false
        end
    end
end


return try_move
